opm/simulators/linalg/bda/rocalutionSolverBackend.cpp

```diff
@@ -0,0 +1,227 @@
+    }
+    if (verbosity >= 3) {
+        std::ostringstream out;
+        out << "rocalutionSolver::convert_matrix(): " << t.stop() << " s";
+        OpmLog::info(out.str());
+    }
+}
+
+
+// copy result to host memory
+// caller must be sure that x is a valid array
+template <unsigned int block_size>
+void rocalutionSolverBackend<block_size>::get_result(double *x) {
+    Timer t;
+
+    std::copy(h_x.begin(), h_x.end(), x);
+
+    if (verbosity >= 3) {
+        std::ostringstream out;
+        out << "rocalutionSolver::get_result(): " << t.stop() << " s";
+        OpmLog::info(out.str());
+    }
+} // end get_result()
+
+
+template <unsigned int block_size>
+SolverStatus rocalutionSolverBackend<block_size>::solve_system(std::shared_ptr<BlockedMatrix> matrix,
+                                                           double *b,
+                                                           [[maybe_unused]] std::shared_ptr<BlockedMatrix> jacMatrix,
+                                                           [[maybe_unused]] WellContributions& wellContribs,
+                                                           BdaResult &res)
+{
+    if (initialized == false) {
+        initialize(matrix);
+    }
+
+    tmp_rowpointers = new int[Nb+1];
+    tmp_colindices = new int[nnzb];
+    tmp_nnzvalues = new double[nnzb*block_size*block_size];
+
+    convert_matrix(matrix);
+
+    rocalution::LocalVector<double> roc_x;
+    rocalution::LocalVector<double> roc_rhs;
+    rocalution::LocalMatrix<double> roc_mat;
+
+    // this also transfers ownership to the allocated memory to rocalution
+    // and sets the tmp_* pointers to nullptr
+    roc_mat.SetDataPtrBCSR(
+        &tmp_rowpointers,
+        &tmp_colindices,
+        &tmp_nnzvalues,
+        "matrix A", nnzb, Nb, Nb, block_size);
+
+    roc_mat.MoveToAccelerator();
+    roc_x.MoveToAccelerator();
+    roc_rhs.MoveToAccelerator();
+
+    roc_x.Allocate("x", roc_mat.GetN());
+    roc_rhs.Allocate("rhs", roc_mat.GetN());
+
+    // initialize vectors
+    roc_rhs.CopyFromData(b);
+    roc_x.Zeros();
+
+    roc_solver->Clear();
+    roc_solver->SetOperator(roc_mat);
+    roc_solver->SetPreconditioner(*roc_prec);
+
+    // the implementation of ILU::ReBuildNumeric() does not exist at the time of writing
+    // so it just calls ILU::Build() everytime
+    roc_solver->ReBuildNumeric();
+
+    // actually solve
+    Dune::Timer t_solve;
+    roc_solver->Solve(roc_rhs, &roc_x);
+
+    // roc_solver->GetSolverStatus() returns:
+    // 0, if no criteria has been reached yet
+    // 1, if absolute tolerance has been reached
+    // 2, if relative tolerance has been reached
+    // 3, if divergence tolerance has been reached
+    // 4, if maximum number of iteration has been reached
+
+    res.elapsed = t_solve.stop();
+    res.iterations = roc_solver->GetIterationCount();
+    res.reduction = roc_solver->GetCurrentResidual();
```

Comment:
It seems impossible to get the initial residual.
It is possible to record all residuals in a vector, but this can only be written to a file.
It is also possible to print the solver status via:
```c++
LOG_INFO("IterationControl RELATIVE criteria has been reached: "
                     << "res norm=" << std::abs(this->current_res_) << "; "
                     << "rel val=" << this->current_res_ / this->initial_residual_ << "; "
                     << "iter=" << this->iteration_);
```
[source file](https://github.com/ROCmSoftwarePlatform/rocALUTION/blob/4780aa22a6eda923e6f9f6ea2c72d478985d1834/src/solvers/iter_ctrl.cpp)

Reply:
You could simply compute it as you should have everything: norm(b-Ax), where x is the initial guess.

Reply:
The initial guess is hardcoded to 0, so I just use `norm(b)` now.